crates/telio-firewall/src/firewall.rs

```diff
@@ -608,15 +644,79 @@ impl StatefullFirewall {
     }
 
     fn handle_inbound_icmp<'a, P: IpPacket<'a>>(&self, peer: &PublicKey, ip: &P) -> bool {
-        let icmp_packet = unwrap_option_or_return!(P::Icmp::new(ip.payload()), false);
+        let icmp_packet = unwrap_option_or_return!(IcmpPacket::new(ip.payload()), false);
 
-        if P::Icmp::BLOCKED_TYPES.contains(&icmp_packet.get_type()) {
+        if P::Icmp::BLOCKED_TYPES.contains(&icmp_packet.get_icmp_type().0) {
             telio_log_trace!("Dropping ICMP packet {:?} {:?}", ip, peer);
             return false;
         }
 
-        telio_log_trace!("Accepting ICMP packet {:?} {:?}", ip, peer);
-        true
+        let mut icmp_cache = unwrap_lock_or_return!(self.icmp.lock(), false);
+        let key = unwrap_option_or_return!(Self::build_icmp_key(ip, true), false);
+        if icmp_cache.get(&key).is_some() {
+            telio_log_trace!("Matched ICMP conntrack entry {:?}", key,);
+            telio_log_trace!("Removing ICMP conntrack entry {:?}", key);
+            icmp_cache.remove(&key);
+            telio_log_trace!("Accepting ICMP packet {:?} {:?}", ip, peer);
+            true
+        } else {
+            false
+        }
+    }
+
+    fn build_icmp_key<'a, P: IpPacket<'a>>(ip: &P, inbound: bool) -> Option<IcmpConn> {
+        let icmp_packet = unwrap_option_or_return!(IcmpPacket::new(ip.payload()), None);
+        let (response_type, has_identifier_sequence) = {
+            use IcmpTypes as v4;
+            use Icmpv6Types as v6;
+            let it = icmp_packet.get_icmp_type().0;
+            // Request messages get the icmp type of the corresponding reply message
+            // Reply messages get their own icmp type, to match that of its corresponding request
+            // Outbound reply messages don't get a key since we're not expecting a response
+            if it == v4::EchoRequest.0 || (it == v4::EchoReply.0 && inbound) {
```

Comment:
I think you have this logic backwards. You don't really care about replies. You care about not letting in requests. So I would imagine having